server/src/main/java/com/ca/lsp/core/cobol/semantics/outline/OutlineTreeBuilder.java

```diff
@@ -0,0 +1,192 @@
+      variables.removeLast();
+    }
+    Variable groupItem = variables.peekLast();
+    if (level == 88) {
+      if (latestVariable == null) {
+        addNodeToParent(outlineNode, parserRuleContext);
+      } else {
+        latestVariable.getChildren().add(outlineNode);
+      }
+      return;
+    }
+    latestVariable = outlineNode;
+    if (groupItem != null) {
+      groupItem.outlineNode.getChildren().add(outlineNode);
+    } else {
+      addNodeToParent(outlineNode, parserRuleContext);
+    }
+    if (NodeType.STRUCT == nodeType ||
+        NodeType.REDEFINES == nodeType) {
+      variables.addLast(new Variable(level, outlineNode));
+    }
+  }
+
+  public OutlineTree build(Multimap<String, Position> copybookUsages) {
+    rootNodes.forEach(this::recalculateStructRange);
+    for (Map.Entry<String, Position> copybook: copybookUsages.entries()) {
+      String name = copybook.getKey();
+      Position position = copybook.getValue();
+      Range outlineRange = new Range(
+          new org.eclipse.lsp4j.Position(position.getLine(), position.getCharPositionInLine()),
+          new org.eclipse.lsp4j.Position(position.getLine(), position.getCharPositionInLine()+name.length())
+      );
+      DocumentSymbol outlineNode = new DocumentSymbol("COPY " + name, NodeType.COPYBOOK.getSymbolKind(), outlineRange, outlineRange, "", List.of());
+      addOutlineNodeToTree(rootNodes, outlineNode);
+    }
+    return new OutlineTree(rootNodes, true);
+  }
+
+  private void recalculateStructRange(DocumentSymbol node) {
+    node.getChildren().forEach(this::recalculateStructRange);
+    if (node.getKind() == NodeType.STRUCT.getSymbolKind()) {
+      node.getRange().setEnd(node.getChildren().stream()
+          .map(DocumentSymbol::getRange)
+          .map(Range::getEnd)
+          .max(Comparator.comparingInt(org.eclipse.lsp4j.Position::getLine)
+              .thenComparingInt(org.eclipse.lsp4j.Position::getCharacter))
+          .orElse(node.getRange().getEnd()));
+    }
+  }
+
+  private void addOutlineNodeToTree(List<DocumentSymbol> outlineNodes, DocumentSymbol outlineNode) {
+    outlineNodes.stream()
+        .filter(it -> RangeUtils.isInsideRange(it, outlineNode))
+        .findFirst()
+        .ifPresentOrElse(
+            it -> addOutlineNodeToTree(it.getChildren(), outlineNode),
+            () -> outlineNodes.add(outlineNode));
+  }
+
+  private DocumentSymbol constructNode(String name, NodeType type, ParserRuleContext parserRuleContext) {
+    Range range = extractRange(parserRuleContext);
+    if (range == null) {
+      return null;
+    }
+    return new DocumentSymbol(name, type.getSymbolKind(), range, range, "", new ArrayList<>());
+  }
+
+  private Range extractRange(ParserRuleContext parserRuleContext) {
+    org.eclipse.lsp4j.Position start = extractPosition(parserRuleContext.start);
+    org.eclipse.lsp4j.Position stop = extractPosition(parserRuleContext.stop);
```

Comment:
In this case, the positions will be null if the rule is adjacent to a copybook. I would recommend a filtration for tokens by channel 0 - they are guaranteed to present in the mapping.  The tokens might be taken from the `BufferedTokenStream` by the source interval

Reply:
There is another restriction in the Outline view. We must show things that belong only to the opened program file.
IDE can't navigate the user to another file from the Outline view. There is no way to open a copybook when the user clicks on some Outline view element. The main idea of that behavior is hiding things that came from copybooks.